Add terraform module to create soaktest workers on GCP

lahsivjar opened the pull request:
## Motivation/summary

Workers will generate continuous and sustained load on the APM-Server to allow for soak testing. Workers will be long running processes with capabilities to recover automatically on failures or generate alert notifications for non-recoverable errors. The terraform script included in this PR makes managing workers easy.

## Checklist

~~- [ ] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)~~
~~- [ ] Update [package changelog.yml](https://github.com/elastic/apm-server/blob/main/apmpackage/apm/changelog.yml) (only if changes to `apmpackage` have been made)~~
- [x] Documentation has been updated

## How to test these changes

As part of the soak test project, the worker should already be running and sending events to a central ESS APM-Server. If that is not the case, the terraform module can be tested by:

```
$ cd testing/apmsoak
$ make apply # this will ask for some configurations related to the target APM Server
```

On success, a worker node should be created sending data to the target APM Server. Assert that APM Server is receiving load.

## Related issues

Closes #8534

apmmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`42/42`) | :green_heart: 
Files | 91.96% (`183/199`) | :+1: 
Classes | 93.435% (`427/457`) | :+1: 
Methods | 89.163% (`1086/1218`) | :+1: 
Lines | 76.886% (`13332/17340`) | :+1: 0.058
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

apmmachine commented:
### :books: Go benchmark report

Diff with the `main` branch

```
name                                                                                           old time/op    new time/op    delta
pkg:github.com/elastic/apm-server/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/decoder goos:linux goarch:amd64
CompressedRequestReader/uncompressed-8                                                            168ns ± 1%     190ns ±11%  +13.09%  (p=0.008 n=5+5)
CompressedRequestReader/gzip_content_encoding-8                                                  5.10µs ± 0%    6.00µs ± 0%  +17.78%  (p=0.008 n=5+5)
CompressedRequestReader/deflate_content_encoding-8                                               5.05µs ± 3%    4.94µs ± 1%   -2.17%  (p=0.008 n=5+5)
pkg:github.com/elastic/apm-server/model/modelindexer goos:linux goarch:amd64
ModelIndexer/DefaultCompression-8                                                                9.17µs ± 1%    9.26µs ± 1%   +1.01%  (p=0.032 n=5+5)
pkg:github.com/elastic/apm-server/processor/stream goos:linux goarch:amd64
BackendProcessor/invalid-event-type.ndjson-8                                                     8.51µs ± 1%    8.34µs ± 1%   -1.97%  (p=0.008 n=5+5)
BackendProcessor/invalid-event.ndjson-8                                                          35.2µs ± 2%    34.7µs ± 1%   -1.35%  (p=0.032 n=5+5)
BackendProcessor/transactions_spans_rum.ndjson-8                                                 18.8µs ± 2%    18.5µs ± 1%   -1.76%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/ratelimit.ndjson-8                   57.9µs ± 1%    59.2µs ± 2%   +2.29%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/transactions-huge_traces.ndjson-8    23.9µs ± 2%    25.2µs ± 3%   +5.13%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4/events.ndjson-8                      45.0µs ± 2%    45.8µs ± 1%   +1.76%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4/transactions-huge_traces.ndjson-8    15.3µs ± 1%    15.0µs ± 1%   -1.91%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/minimal-service.ndjson-8             3.52µs ± 1%    3.61µs ± 2%   +2.53%  (p=0.032 n=5+5)
pkg:github.com/elastic/apm-server/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
TraceGroups-8                                                                                     197ns ± 5%     186ns ± 3%   -5.43%  (p=0.032 n=5+5)
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
ReadEvents/json_codec/199_events-8                                                               2.51ms ± 1%    2.48ms ± 1%   -1.42%  (p=0.032 n=5+5)

name                                                                                           old alloc/op   new alloc/op   delta
pkg:github.com/elastic/apm-server/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/decoder goos:linux goarch:amd64
CompressedRequestReader/gzip_content_encoding-8                                                  45.5kB ± 0%    45.5kB ± 0%   +0.02%  (p=0.008 n=5+5)
pkg:github.com/elastic/apm-server/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/processor/stream goos:linux goarch:amd64
BackendProcessor/heavy.ndjson-8                                                                  13.7MB ± 0%    13.7MB ± 0%   -0.24%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel0/transactions_spans_rum_2.ndjson-8    12.3kB ± 1%    12.4kB ± 1%   +0.95%  (p=0.024 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/optional-timestamps.ndjson-8         21.8kB ± 1%    22.0kB ± 1%   +1.02%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/spans.ndjson-8                       92.9kB ± 0%    92.6kB ± 0%   -0.40%  (p=0.016 n=5+4)
pkg:github.com/elastic/apm-server/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64
ReadEvents/json_codec_big_tx/1_events-8                                                          10.2kB ± 0%    10.2kB ± 0%   +0.02%  (p=0.008 n=5+5)

name                                                                                           old allocs/op  new allocs/op  delta
pkg:github.com/elastic/apm-server/agentcfg goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/decoder goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/model/modelindexer goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/processor/stream goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/publish goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/spanmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/aggregation/txmetrics goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling goos:linux goarch:amd64
pkg:github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage goos:linux goarch:amd64

name                                                                                           old speed      new speed      delta
pkg:github.com/elastic/apm-server/processor/stream goos:linux goarch:amd64
BackendProcessor/invalid-event-type.ndjson-8                                                   45.9MB/s ± 1%  46.9MB/s ± 1%   +2.01%  (p=0.008 n=5+5)
BackendProcessor/invalid-event.ndjson-8                                                        21.8MB/s ± 2%  22.1MB/s ± 1%   +1.36%  (p=0.024 n=5+5)
BackendProcessor/transactions_spans_rum.ndjson-8                                               61.5MB/s ± 2%  62.6MB/s ± 1%   +1.79%  (p=0.024 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/ratelimit.ndjson-8                 72.8MB/s ± 1%  71.2MB/s ± 2%   -2.23%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel2/transactions-huge_traces.ndjson-8   132MB/s ± 2%   126MB/s ± 3%   -4.86%  (p=0.008 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4/events.ndjson-8                     165MB/s ± 2%   162MB/s ± 1%   -1.73%  (p=0.032 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel4/transactions-huge_traces.ndjson-8   207MB/s ± 1%   211MB/s ± 1%   +1.95%  (p=0.016 n=5+5)
BackendProcessorParallel/BenchmarkBackendProcessorParallel8/minimal-service.ndjson-8            121MB/s ± 1%   118MB/s ± 2%   -2.44%  (p=0.032 n=5+5)

```

_report generated with https://pkg.go.dev/golang.org/x/perf/cmd/benchstat_
<!--COMMENT_GENERATED_WITH_ID_bench.diff-->

marclop commented on testing/apmsoak/Makefile:
```diff
@@ -0,0 +1,34 @@
+.default: all
+
+.PHONY: all
+all: $(SSH_KEY) terraform.tfvars init apmsoak apply
```
I don't think there's any `SSH_KEY` target here, shouldn't we declare a default value for `SSH_KEY` which points to `~/.ssh/google_compute_engine`?

marclop commented on testing/infra/terraform/modules/soaktest_workers/README.md:
```diff
@@ -0,0 +1,58 @@
+## Pre-requisites
+
+1. Follow the steps in [this guide](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/getting_started).
+2. The project uses [OS login](https://cloud.google.com/compute/docs/oslogin) to ssh into the created worker node, make sure that your google account has the [required privileges](https://cloud.google.com/compute/docs/oslogin/set-up-oslogin#grant-iam-roles)
```
Will this work out of the box for our accounts?

lahsivjar replied:
Yes, since we have editor access on elastic-apm project

marclop commented on testing/infra/terraform/modules/soaktest_workers/main.tf:
```diff
@@ -0,0 +1,141 @@
+
+resource "google_compute_firewall" "allow_ssh" {
+  name    = "apmserver-soaktest-allowssh"
+  network = local.network_name
+
+  depends_on = [
+    google_compute_network.apmsoak_worker_network
+  ]
```
We can remove this `depends_on` block if we use the `google_compute_network.apmsoak_worker_network.self_link`. See https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_firewall#network and https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/compute_network#self_link.

```suggestion
  network = google_compute_network.apmsoak_worker_network.self_link
```

lahsivjar replied:
Thanks, sounds cleaner

lahsivjar replied:
Fixed

marclop commented on testing/infra/terraform/modules/soaktest_workers/main.tf:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  network_interface {
+    network = local.network_name
+    access_config {}
+  }
+
+  depends_on = [
+    google_compute_firewall.allow_ssh,
+  ]
```
Same here

lahsivjar replied:
Didn't get this one. While provisioning we require firewall to be created however, we don't actually use the firewall.

marclop replied:
Oh, I see. Ignore this comment then

marclop reviewed (commented):
Mostly looks good!

lahsivjar commented:
@marclop I have updated the PR to remove OSLogin as it was not serving our use-case. I have also updated to use terraform workspaces to better manage production deployment of workers. (apologies for the force push)

marclop commented on testing/apmsoak/main.tf:
```diff
@@ -0,0 +1,32 @@
+      version = ">=4.27.0"
+    }
+  }
+  backend "gcs" {
+    bucket = "apmsoak-tf"
+    prefix = "/tfstate"
+  }
```
I haven't tested using different workspaces with the GCS state set. Did you check it works well?

lahsivjar replied:
Yes, tested with 2 workspaces as well as with a service account

marclop commented on testing/infra/terraform/modules/soaktest_workers/README.md:
```diff
@@ -0,0 +1,52 @@
+## Soaktest worker module
+
+This module sets up worker with load generation binary for soaktest configured as a systemd unit along with required monitoring setup. The workers generated are immutable in nature and are recreated for every configuration change.
```
Aren't we missing the `header.md` file in this PR?

lahsivjar replied:
Sorry, I am not sure where the requirement of `header.md` comes from. Is it required as part of a terraform module?

marclop replied:
How did you generate this `README.md` file?

In the makefile, there is a target called `tf-docs`. When running it, it will use the contents of the `header.md file to generate the `README.md` file. If that file isn't checked in, then the contents will be re-generated in the future and we'll lose this information.

https://github.com/elastic/apm-server/blob/3de716555254975194ad5e6729e3cbf5b4536b88/Makefile#L129-L133

lahsivjar replied:
Ah I see. I used the [terraform-docs cli](https://github.com/terraform-docs/terraform-docs) directly, wasn't aware about the Makefile. Let me update.

marclop commented on testing/infra/terraform/modules/soaktest_workers/main.tf:
```diff
@@ -0,0 +1,141 @@
+  ))
+}
+
+resource "tls_private_key" "worker_login" {
```
I didn't know this resource type existed. How would this work when / if we need to ssh to a remote machine and debug failing `apmsoak` systemd services?

lahsivjar replied:
You can get the private key from the backend state and ssh.

marclop commented on testing/infra/terraform/modules/soaktest_workers/main.tf:
```diff
@@ -0,0 +1,141 @@
+  }
+
+  network_interface {
+    network = local.network_name
```
shouldn't we be using `google_compute_network.worker.self_link` here?

lahsivjar replied:
Good catch, name and self_link both works but self_link will be better. Will update this.

lahsivjar commented on testing/apmsoak/README.md:
```diff
@@ -0,0 +1,53 @@
+## Pre-requisites
+
+1. Install terraform
+2. Make sure that you have access to the GCP project and have enough privileges to create and manage instances. The workign set of required roles to be able to create worker using this module are:
```
Updated the document, fixed typos. I am using the singular `worker` because the module can only create a single worker at once. Let me know if the document makes better sense this way.

marclop commented on testing/apmsoak/README.md:
```diff
@@ -0,0 +1,54 @@
+
+**NOTE: This is unlikely to be useful for non-Elastic employees**
+
+This folder provides tools to enable creation of a worker instance to send load to any APM Server. Worker run the `apmsoak` binary to generate a continuous load and since it is a long running process, it is run as systemd unit configured to restart on failure. Monitoring is also setup for the worker as part of the terraform module to monitor the node resources and the systemd unit status.
```
```suggestion
This folder provides tools to enable creation of a worker instance to send load to any APM Server. The worker runs the `apmsoak` binary to generate a continuous load and since it is a long running process, it is run as systemd unit configured to restart on failure. Monitoring is also setup for the worker as part of the terraform module to monitor the node resources and the systemd unit status.
```

marclop reviewed (approved):
LGTM!
